File: Java-Core/src/main/java/prompto/value/CharacterRange.java
package prompto.value;

import prompto.intrinsic.PromptoRange;
import prompto.type.CharacterType;


public class CharacterRange extends RangeBase<Character> {

	static class PromptoCharacterRange extends PromptoRange<Character> {

		public PromptoCharacterRange(prompto.value.Character low, prompto.value.Character high) {
			super(low, high);
		}
		
		@Override
		public prompto.value.Character getItem(long item) {
			char result = (char)(low.getValue() + item - 1);
			if(result>high.getValue())
				throw new IndexOutOfBoundsException();
			return new prompto.value.Character(result);
		}
		
		@Override
		public PromptoCharacterRange slice(long first, long last) {
			last = adjustLastSliceIndex(last);
			return new PromptoCharacterRange(getItem(first), getItem(last));
		}
	
		@Override
		public long length() {
			return 1L + high.getValue() - low.getValue();
		}

	}
	
	public CharacterRange(Character left, Character right) {
		this(new PromptoCharacterRange(left, right));
	}
	
	public CharacterRange(PromptoRange<Character> range) {
		super(CharacterType.instance(), range);
	}
	
	@Override
	public int compare(Character o1, Character o2) {
		return o1.compareTo(o2);
	}
	
	@Override
	public RangeBase<Character> newInstance(PromptoRange<Character> range) {
		return new CharacterRange(range);
	}


}

File: Java-CodeServer/src/main/resources/js/prompto/value/CharacterRange.js
var Range = require("./Range").Range;
var Integer = require("./Integer").Integer;
var Character = null;
var CharacterType = null;

exports.resolve = function() {
    Character = require("./Character").Character;
    CharacterType = require("../type/CharacterType").CharacterType;
}

function CharacterRange(left, right) {
	Range.call(this, CharacterType.instance, left, right);
	return this;
}

CharacterRange.prototype = Object.create(Range.prototype);
CharacterRange.prototype.constructor = CharacterRange;


CharacterRange.prototype.size = function() {
	return 1 + this.high.value.charCodeAt(0) - this.low.value.charCodeAt(0);
};


CharacterRange.prototype.getItem = function(index) {
	var result = this.low.value.charCodeAt(0) + index - 1;
	if(result>this.high.value.charCodeAt(0)) {
		throw new IndexOutOfBoundsException();
	} else {
		return new Character(String.fromCharCode(result));
	}
};

/*
@Override
public Range<Character> newInstance(Character left, Character right) {
	return new CharacterRange(left, right);
}


*/

exports.CharacterRange = CharacterRange;